docs/SCHEMA.md

```diff
@@ -0,0 +1,348 @@
+# .env File Schema
+
+`configure-env` supports any valid `.env` file that can be parsed by libraries like [`dotenv`](https://npm.im/dotenv).
+
+That means it supports:
+
+- "variable declaration": key value pairs in the pattern: `KEY=VALUE`
+- empty declarations are valid but **must** contain a `=`. Example: `KEY=`
+- one key value pair per line
+- empty lines are ignored
+- lines beginning with # are treated as comments
+
+Based on those rules, [special comment annotations](#special-comment-annotations) and [best effort guessing](#best-effort-guessing), this tool will extract meta information for each key value pair that can either be extracted using the `parse` or `parseFile` function or that will be used by the CLI component of `configure-env` to prompt users for input of said environment variables.
+
+## Special Comment Annotations
+
+Any comments that are immediately before a "variable declaration" will be evaluated in the context of the variable declaration. Special comments can be used to define:
+
+- [`required`](#required)
+- [`format`](#format)
+- [`description`](#description)
+- [`link`](#link)
+- [`default`](#default)
+- [`configurable`](#configurable)
+
+These comments can be used in any order as long as they are right before the variable declaration with no empty line before it.
+
+### `required`
+
+Using `# required: true` or `# required: false` can define whether a variable has to be configured in order for the app to work.
+
+**Default Value**: `true`
+
+**Example `.env` file:**
+
+```bash
+# required: true
+TWILIO_ACCOUNT_SID=
+```
+
+**Parsed Result (with omitted default values):**
+
+```json
+[{ "key": "TWILIO_ACCOUNT_SID", "required": true }]
+```
+
+### `format`
+
+A format comment can be used to inform a user what types of value is being expected for a variable.
+
+Valid values are:
+
+- `text` (plain text)
+- `phone_number` (representing the [E.164 Phone Number Format](https://www.twilio.com/docs/glossary/what-e164))
+- `email`
+- `url` (full or relative path)
+- `sid` ([Twilio specific resource identifier format](https://www.twilio.com/docs/glossary/what-is-a-sid))
+- `integer` (specifically any whole number)
+- `number` (any valid number including floating point numbers with `.` as floating point indicator)
+- `secret` (the same as text but UIs might decide to hide the user input)
+- `list(<X>)` (a comma separated list of values, where `<X>` defines any of the formats above). Example: `list(email)` for `support@twilio.com,open-source@twilio.com`. Use `text` for any type
+- `nested_list(<X>,<Y>)` (a comma and semicolon separted list. often used for key value pairs. `<X>` and `<Y>` represent the format for the first any any following entry). Example: `nested_list(email,phone_number)` for: `help@twilio.com,+1222333444;support@twilio.com,+13334445555`
```

Comment:
The common use case here is to provide a rudimentary set of credentials for example. Say you want to have multiple users instead of one set of variables.

Reply:
what if it was just `list` but you can infer the additional functionality by virtue of more than 1 type specified to the list args? You could probably keep the code as is in with a simple additional check on the args size to determine how to handle it.

Reply:
I think it's a bit more explicit for users if we can inform them about it being list vs a nested list. I could go that way though. @bharatmurali what do you think?

Reply:
we'll rename it to `map`